---
title: "SingleCell of pancreas tissue"
author: "PopovMD"
date: "2025-09-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## SingleCell data analysis

I was given these files [FACS data from mice tissues]:  
<https://figshare.com/articles/dataset/Single-cell_RNA-seq_data_from_Smart-seq2_sequencing_of_FACS_sorted_cells/5715040?file=10039267>   
to perform these sequence of actions [SingleCell analysis]:  
<https://docs.google.com/document/d/1FUbsBUp7Kqh9BeqQ81wVuHGu4mAc6P8mC6l4VR75LKk/edit?tab=t.0#heading=h.ghk5t6a6l7dx>.

Script files could be found in the same folder, where you found this .Rmd.  

After downloading all necessary files, googling some information about the biological methods and choosing tissue with two subtissues (pancreas), I hopped right to the analysis pipeline.

### Creating Seurat object

```{r libraries, results='hide', message=FALSE, warning=FALSE}
library(dplyr)
library(Seurat)
library(patchwork)
library(Matrix)
```

```{r CreateSeuratObject, message=FALSE, warning=FALSE}
# creating a Seurat Object out of csv table of raw counts for pancreas tissue
raw_counts <- read.csv(file = "./pancreas.raw.counts/Pancreas-counts.csv",
                       sep = ",", header = TRUE, row.names = 1)
colnames(raw_counts) <- gsub("\\.", '_', colnames(raw_counts))
sparse_counts <- as.sparse(as.matrix(raw_counts))
pancr <- CreateSeuratObject(counts = sparse_counts,
                            min.cells = 3, 
                            min.features = 200,
                            names.field = 5, # creating clusters by gender: M and F
                            project = "pancreas")
```

```{r head metadata}
head(pancr@meta.data, 10)
```

  
### QC and cells selection

```{r QC1, message=FALSE, warning=FALSE}
# adding a new column to the meta.data, that contains the persantage of reads 
# that map to the mitochondrial genome (by using [[ ]]).
pancr[["percent.mt"]] <- PercentageFeatureSet(pancr, pattern = "^MT-")
head(pancr@meta.data, 5)

plot1 <- FeatureScatter(pancr, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pancr, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

sum(pancr@meta.data$percent.mt != 0) # = 0
```


There are no mitochondrial genes at all, so I carried the same procedures over other several datasets, resulting for them to be in the same condition. This is possible due to two independent situations: mithochondrial genes were discarded in advance, or maybe this not in fact a Single Cell, but "Single Nucleus" (method to sequence only nucleus genes).


```{r QC2, message=FALSE, warning=FALSE}
# removing samples with less than 200 genes and more that 7500:
pancr <- subset(pancr, subset = nFeature_RNA > 200 & nFeature_RNA < 7500)
# maybe it makes sense to remove samples with too many reads

FeatureScatter(pancr, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```



### Normalizing

```{r Normalization, message=FALSE, warning=FALSE}
pancr_norm <- NormalizeData(pancr, normalization.method = "LogNormalize", scale.factor = 10000)
```


LogNormalize method: normalizing by total expression, multiplying by scale.factor and then log-transforms the result.  
Assumption: each cell originally contains the same number of RNA molecules.  


### Selection of highly variable features

```{r selection, message=FALSE, warning=FALSE}
pancr_norm_varfeat <- FindVariableFeatures(pancr_norm, 
                                           selection.method = "vst", 
                                           nfeatures = 3000)
plot1 <- VariableFeaturePlot(pancr_norm_varfeat)

top10 <- head(VariableFeatures(pancr_norm_varfeat), 5)
LabelPoints(plot = plot1, points = top10, repel = TRUE)
```



### Linear transformation of the data:

```{r scaling, message=FALSE, warning=FALSE}
all.genes <- rownames(pancr_norm_varfeat)
# transforming all of the genes, not only variable
pancr_linear <- ScaleData(pancr_norm_varfeat, features = all.genes)
```


### Performing linear dimensional reduction

Constructing PCA.  

```{r PCA, message=FALSE, warning=FALSE}
# using only variable features, determined previously (by default)
pancr_linear <- RunPCA(pancr_linear)
print(pancr_linear[['pca']], nfeatures = 6)
```

For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings (correlation coefficients), representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset.  

```{r pca plot1, message=FALSE, warning=FALSE}
VizDimLoadings(pancr_linear, dims = 1:2, reduction = "pca")
```

The above graph shows, that the data is not really heterogeneous.

```{r pca plot2, message=FALSE, warning=FALSE}
DimPlot(pancr_linear, reduction = "pca") + NoLegend()
```

By this standard PCA picture we can appreciate, that the difference in transcription products for males and females is not that big.  

```{r heatmap1, message=FALSE, warning=FALSE}
DimHeatmap(pancr_linear, dims = 1:6, cells = 500, balanced = TRUE)
```

And this plot gives us an opportunity to find out, which genes are responsible the most for defining the first 6 principal components.

```{r how many pcs, message=FALSE, warning=FALSE}
# Choosing the number of principal components by percentage of variance explained by each one
ElbowPlot(pancr_linear, ndims = 50)
```

I would say I see the rightest elbow at the point of 23rd PC.


### Clustering the cells

Providing automated clusterization for further needs (using first 23 PCs):

```{r clusterization, message=FALSE, warning=FALSE}
pancr_clust23 <- FindNeighbors(pancr_linear, dims = 1:23)
pancr_clust23 <- FindClusters(pancr_clust23, resolution = 0.4)

head(Idents(pancr_clust23))
```


### Non-linear dimensional reduction

```{r UMAPs, message=FALSE, warning=FALSE}
# UMAP for male-female
pancr_linear <- RunUMAP(pancr_linear, dims = 1:23)
umap_M_F_plot <- DimPlot(pancr_linear, 
        reduction = "umap")
        
# UMAP for defined clusters
pancr_clust23 <- RunUMAP(pancr_clust23, dims = 1:23)
umap_clusters_plot <- DimPlot(pancr_clust23, 
        reduction = "umap",
        label = TRUE)
  
umap_clusters_plot + umap_M_F_plot
```


### Creating the bar plot

```{r barplot, message=FALSE, warning=FALSE}
library(ggplot2)

# Bar plot
ggplot(pancr_clust23@meta.data, 
       aes(x = Idents(pancr_clust23), fill = pancr_clust23@meta.data$orig.ident)) +
       geom_bar(position = "fill") +
       labs(x = "Cluster", y = "Proportion", fill = "Gender") +
  theme_minimal()
```


BarPlot clearly gives as the understanding of which clusters have bigger shift to a specific gender of mice: the second is the most "feminine" one [>0.75], and the sixth cluster stands for the male representatives [almost 1].  
The representation of these facts is also can be seen on UMAP-plots above. Furthermore, we can say that for all other clusters the percentage of both genders are less shifted, and no clusters consist only of one gender (but the best candidate for this - the 6th cluster).


### Analysing subtissues

For inserting information about subtissues from metadata file into the count.csv table, I wrote a small python-code. You can find it in the same folder.

```{r subtissues, include=FALSE}
# creating a Seurat Object out of csv table of raw counts for pancreas tissue
raw_counts <- read.csv(file = "/Users/svetlana/Desktop/SingleCell_test-task_for_lab/pancreas.raw.counts/pancreas_renamed_subtissue.csv",
                       sep = ",", header = TRUE, row.names = 1)
# making identity classes - by subtissues
colnames(raw_counts) <- gsub("\\.", '_', colnames(raw_counts))
sparse_counts <- as.sparse(as.matrix(raw_counts))
pancr <- CreateSeuratObject(counts = sparse_counts,
                            min.cells = 3, 
                            min.features = 200,
                            names.field = 8, # making identity classes - by subtissues
                            project = "pancreas")


## QC and cell selection

# removing samples with less than 200 genes and more that 7500:
pancr <- subset(pancr, subset = nFeature_RNA > 200 & nFeature_RNA < 7500)

scatter <- FeatureScatter(pancr, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")


## Normalizing

pancr_norm <- NormalizeData(pancr, normalization.method = "LogNormalize", scale.factor = 10000)


pancr_norm_varfeat <- FindVariableFeatures(pancr_norm, 
                                           selection.method = "vst", 
                                           nfeatures = 3000)


## Linear transformation of the data:

all.genes <- rownames(pancr_norm_varfeat)
pancr_linear <- ScaleData(pancr_norm_varfeat, features = all.genes)



## Performing linear dimensional reduction

# using only variable features, determined previously (by default)
pancr_linear <- RunPCA(pancr_linear)
#print(pancr_linear[['pca']], nfeatures = 6)

#VizDimLoadings(pancr_linear, dims = 1:2, reduction = "pca")

PCA <- DimPlot(pancr_linear, reduction = "pca") + NoLegend()

#DimHeatmap(pancr_linear, dims = 1, cells = 500, balanced = TRUE)

#ElbowPlot(pancr_linear, ndims = 50)


## Clustering the cells

pancr_clust23_sub <- FindNeighbors(pancr_linear, dims = 1:23)
pancr_clust23_sub <- FindClusters(pancr_clust23_sub, resolution = 0.4)


pancr_linear <- RunUMAP(pancr_linear, dims = 1:23)

umap_M_F_plot <- DimPlot(pancr_linear, 
                         reduction = "umap")
#
```
```{r subtissues plots, echo=FALSE, message=FALSE, warning=FALSE}
scatter 
PCA + umap_M_F_plot
```
```{r barplot2, echo=FALSE, message=FALSE, warning=FALSE}
ggplot(pancr_clust23_sub@meta.data, 
       aes(x = Idents(pancr_clust23_sub), fill = pancr_clust23_sub@meta.data$orig.ident)) +
  geom_bar(position = "fill") +
  labs(x = "Cluster", y = "Proportion", fill = "Type") +
  theme_minimal()
```

The situation we are witnessing on the plots above is dramatically different from the previous one: almost each cluster consists only from one type of cells, endocrine or exocrine (you can clearly see it thanks to barplot). Clusters 2,4,5 are dedicated to the exocrine subtissue, clusters 0,1,3,6,8 - to the endocrine. Also, there are two clusters with nearly equlibrium percentage for subtissues.  
And you can see the strict clusterization on the UMAP-plot, that is provided.



```{r diffexp genes, message=FALSE, warning=FALSE}
# find all markers (of DE of genes) distinguishing Exocrine class from Endocrine one
clusters.markers <- FindMarkers(pancr_linear, 
                                ident.1 = 'Endocrine', 
                                ident.2 = 'Exocrine')

clusters.markers %>%
  dplyr::filter(avg_log2FC > 1) %>%
  slice_head(n = 60) %>%
  ungroup() -> top60

# HeatMap
DoHeatmap(pancr_linear, features = row.names(top60)) + NoLegend()
```

Unfortunately, we can not clearly see the separation of two subtissues on the heatmap: it seems like exocrine one is merely above zero rate of expression, however, this plot is not full: it represents only top 60 genes by the expression rate. So, we unintentionally chose only whose genes, which are transcribed mostly in endocrine subtissue.  
Nevertheless, we still can confirm, that by the expression rate of these genes, these subtissues are different.







